WordPress/Classes/Utility/BackgroundTasks/WeeklyRoundupBackgroundTask.swift

```diff
@@ -34,89 +35,154 @@ private class WeeklyRoundupDataProvider {
-        endDateComponents.weekday = 1
-
-        // The DateComponents timezone is ignored when calling `Calendar.current.nextDate(...)`, so we need to
-        // create a GMT Calendar to perform the date search using it, instead.
-        var gmtCalendar = Calendar(identifier: .gregorian)
-        gmtCalendar.timeZone = TimeZone(secondsFromGMT: 0)!
-
-        guard let periodEndDate = gmtCalendar.nextDate(after: Date(), matching: endDateComponents, matchingPolicy: .nextTime, direction: .backward) else {
-            DDLogError("Something's wrong with the preiod end date selection.")
+    /// Fetches the top site statistics for a given array of blogs.
+    ///
+    /// This method asynchronously fetches weekly statistics for each blog in the provided list.
+    /// After all data has been fetched, it filters out the top 5 sites based on the received stats
+    /// and returns the result through the provided completion handler.
+    ///
+    /// **Note:** The completion handler is executed on the context's queue.
+    private func getTopSiteStats(
+        from sites: [Blog],
+        in context: NSManagedObjectContext,
+        completion: @escaping (Result<SiteStats?, Error>) -> Void
+    ) {
+        guard let periodEndDate = getPeriodEndDate() else {
+            DDLogError("Something's wrong with the period end date selection.")
             return
         }
 
-        var blogStats = [Blog: StatsSummaryData]()
-        var statsProcessed = 0
+        let group = DispatchGroup()
+        var blogStats = SiteStats()
 
         for site in sites {
             guard let authToken = site.account?.authToken else {
                 continue
             }
 
-            let wpApi = WordPressComRestApi.defaultApi(oAuthToken: authToken, userAgent: WPUserAgent.wordPress())
-
-            guard let dotComID = site.dotComID?.intValue else {
-                onError(DataRequestError.dotComSiteWithoutDotComID(site))
-                continue
-            }
-
-            let statsServiceRemote = StatsServiceRemoteV2(wordPressComRestApi: wpApi, siteID: dotComID, siteTimezone: site.timeZone)
+            group.enter()
 
-            statsServiceRemote.getData(for: .week, endingOn: periodEndDate, limit: 1) { (timeStats: StatsSummaryTimeIntervalData?, error) in
+            self.fetchStats(for: site, endingOn: periodEndDate, authToken: authToken, in: context) { [weak self] result in
                 defer {
-                    statsProcessed = statsProcessed + 1
-
-                    if statsProcessed == sites.count {
-                        let bestBlogStats = self.filterBest(5, from: blogStats)
-
-                        completion(.success(bestBlogStats))
+                    group.leave()
+                }
+                guard let self else {
+                    return
+                }
+                switch result {
+                case .failure(let error):
+                    self.onError(error)
```

Comment:
I'm a little bit confused by this function. When an error occurs, should `onError` be called, or should `completion` block be called, or maybe both? Also, ifHTTP API calls fail for all sites (or more than one site), this `onError` closure will be called multiple times. Is that expected?

Reply:
This `onError` closure is inherited from the previous implementation. And honestly I find having both `onError` and a `completionHandler` a bit confusing.

This `onError` closure comes from the block we pass to `WeeklyRoundupBackgroundTask.run` method. See the following screenshot.

<img src="https://github.com/wordpress-mobile/WordPress-iOS/assets/9609223/ed05eab8-6573-4f08-a722-698111e34649" width=400 />

The difference between `onError` and `completionHandler` is that `onError` is called when the task fails to fetch the stats of a site. Instead of interrupting the whole task because of a single site failure, the error is reported through that `onError` and the task continues.